D103-D106 helper routines ffde_* (QDYNAM dump writers; units 1140 in, 1141 checkpoint stream, 1142 q0, 1143 ain/out, 1144 fin, 1149 geometry; ffd_qdyn_* files). Generated by diff -u against the PRISTINE ATM_DRV.f (2026-10-05); inserted after 'end subroutine new_io_atmvars' (unchanged code); applies to the pristine file or after other ATM_DRV_*.patch files. Uses existing modules only (model_com, resolution, atm_com, somtq_com, geom, quscom, diag_com, constant). Env FFD_START/FFD_NSTEP, optional FFD_QDYN_STRESS=<factor> and FFD_QDYN_COMP=<subset of uvw>.

--- ATM_DRV.f.orig
+++ ATM_DRV.f
@@ -1001,6 +1001,347 @@
       return
       end subroutine new_io_atmvars
 
+      logical function ffde_on()
+!@sum full-fidelity-port instrumentation (D103-D106, QDYNAM): true when
+!@+   env FFD_START/FFD_NSTEP select this itime.  Units 1140-1149.
+      use model_com, only : itime
+      implicit none
+      integer, save :: it0=-1, nst=0, init=0
+      integer :: st
+      character(len=64) :: buf
+      if (init.eq.0) then
+        init=1
+        call get_environment_variable('FFD_START',buf,status=st)
+        if (st.eq.0) read(buf,*) it0
+        call get_environment_variable('FFD_NSTEP',buf,status=st)
+        if (st.eq.0) read(buf,*) nst
+      endif
+      ffde_on = .not.(it0.lt.0 .or. itime.lt.it0
+     &     .or. itime.ge.it0+nst)
+      end function ffde_on
+
+      subroutine ffde_open(fu,ph)
+!@sum open ffd_qdyn_<itime>_<ph>.bin (prefix ffd_qdynS<f>_ in a stress run)
+      use model_com, only : itime
+      implicit none
+      integer :: fu
+      character(len=*) :: ph
+      character(len=200) :: fname
+      integer :: ffde_act
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      if (ffde_sfac.eq.1d0) then
+        write(fname,'(A,I0,A,A,A)') 'ffd_qdyn_',itime,'_',trim(ph),
+     &       '.bin'
+      elseif (ffde_comp.eq.'uvw') then
+        write(fname,'(A,I0,A,I0,A,A,A)') 'ffd_qdynS',
+     &       nint(ffde_sfac),'_',itime,'_',trim(ph),'.bin'
+      else
+        write(fname,'(A,I0,A,A,I0,A,A,A)') 'ffd_qdynS',
+     &       nint(ffde_sfac),trim(ffde_comp),'_',itime,'_',trim(ph),
+     &       '.bin'
+      endif
+      open(fu,file=trim(fname),form='unformatted',access='stream',
+     &     status='replace')
+      end subroutine ffde_open
+
+      subroutine ffde_in()
+!@sum QDYNAM inputs (called after MB=MAOLD*KG2MB*AXYP, before AADVQ0).
+!@+   ffd_qdyn_<itime>_in.bin = [itime,sfac,0,0], Q (im,jm,lm),
+!@+   QMOM (9,im,jm,lm), MAOLD (lm,im,jm), MUs,MVs,MWs,MB (im,jm,lm).
+!@+   STRESS TEST (off unless env FFD_QDYN_STRESS=<f>): MUs,MVs,MWs
+!@+   are scaled by f before AADVQ0 (module workspace of this step) so
+!@+   that ncyc>1, ncycxy>1, z_extra branches execute.  Env
+!@+   FFD_QDYN_COMP=<subset of uvw> (default uvw) selects which of
+!@+   MUs,MVs,MWs are scaled; files then get prefix ffd_qdynS<f><comp>_.
+!@+   Also opens the checkpoint stream (unit 1141).
+      use model_com, only : itime
+      use resolution, only : im,jm,lm
+      use atm_com, only : q,maold,mus,mvs,mws,mb
+      use somtq_com, only : qmom
+      use geom, only : axyp,imaxj
+      implicit none
+      logical, external :: ffde_on
+      integer :: fu,st
+      integer :: ffde_act
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      integer, save :: init=0
+      character(len=64) :: buf
+      if (init.eq.0) then
+        init=1
+        ffde_sfac=1d0 ; ffde_act=0 ; ffde_comp='uvw'
+        call get_environment_variable('FFD_QDYN_STRESS',buf,status=st)
+        if (st.eq.0) read(buf,*) ffde_sfac
+        call get_environment_variable('FFD_QDYN_COMP',buf,status=st)
+        if (st.eq.0) ffde_comp=buf(1:3)
+      endif
+      ffde_act=0
+      if (.not.ffde_on()) return
+      ffde_act=1
+      if (ffde_sfac.ne.1d0) then
+        if (index(ffde_comp,'u').gt.0) mus(:,:,:)=mus(:,:,:)*ffde_sfac
+        if (index(ffde_comp,'v').gt.0) mvs(:,:,:)=mvs(:,:,:)*ffde_sfac
+        if (index(ffde_comp,'w').gt.0) mws(:,:,:)=mws(:,:,:)*ffde_sfac
+      endif
+      fu=1140
+      call ffde_open(fu,'in')
+      write(fu) dble(itime),ffde_sfac,0d0,0d0
+      write(fu) q(1:im,1:jm,1:lm)
+      write(fu) qmom(1:9,1:im,1:jm,1:lm)
+      write(fu) maold(1:lm,1:im,1:jm)
+      write(fu) mus(1:im,1:jm,1:lm)
+      write(fu) mvs(1:im,1:jm,1:lm)
+      write(fu) mws(1:im,1:jm,1:lm)
+      write(fu) mb(1:im,1:jm,1:lm)
+      close(fu)
+      fu=1141
+      call ffde_open(fu,'ck')
+      end subroutine ffde_in
+
+      subroutine ffde_geom()
+!@sum one-time geometry/constants: ffd_qdyn_geom.bin = [im,jm,lm,kg2mb,
+!@+   byim(GEOM),byim(QUSCOM),byim(DIAG_COM)], AXYP (im,jm), IMAXJ (jm)
+      use resolution, only : im,jm,lm
+      use geom, only : axyp,imaxj,gbyim=>byim
+      use quscom, only : qbyim=>byim
+      use diag_com, only : dbyim=>byim
+      use constant, only : kg2mb
+      implicit none
+      integer :: fu
+      fu=1149
+      open(fu,file='ffd_qdyn_geom.bin',form='unformatted',
+     &     access='stream',status='replace')
+      write(fu) dble(im),dble(jm),dble(lm),kg2mb,gbyim,qbyim,dbyim
+      write(fu) axyp(1:im,1:jm)
+      write(fu) dble(imaxj(1:jm))
+      close(fu)
+      end subroutine ffde_geom
+
+      subroutine ffde_q0(ncyc,do_z_extra,ncycxy,nstepx,nstepz_extra,
+     &     lminzij,lmaxzij,mw_extra,pv_south,ni_y,ni_z,i_y,i_z)
+!@sum AADVQ0 outputs, ffd_qdyn_<itime>_q0.bin = [itime,ncyc,do_z_extra,
+!@+   0], NCYCXY (lm), NSTEPX (jm,lm), NSTEPZ_EXTRA, LMINZIJ, LMAXZIJ
+!@+   (im,jm), MW_EXTRA (im,jm,lm; raw memory where undefined),
+!@+   MUs,MVs,MWs (im,jm,lm; AFTER AADVQ0), PV_SOUTH (im,lm),
+!@+   NI_CHECKFOBS_Y, NI_CHECKFOBS_Z (jm,lm), then n_y, n_z and
+!@+   triplets (j,l,i) of the i_checkfobs_y / _z lists.
+      use model_com, only : itime
+      use resolution, only : im,jm,lm
+      use atm_com, only : mus,mvs,mws
+      implicit none
+      integer :: ncyc,ncycxy(lm),nstepx(jm,lm),nstepz_extra(im,jm)
+      integer :: lminzij(im,jm),lmaxzij(im,jm)
+      logical :: do_z_extra
+      real*8 :: mw_extra(im,jm,lm),pv_south(im,lm)
+      integer :: ni_y(jm,lm),ni_z(jm,lm),i_y(im,jm,lm),i_z(im,jm,lm)
+      integer :: fu,j,l,k,n
+      integer :: ffde_act,idz
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      integer, save :: ginit=0
+      if (ffde_act.eq.0) return
+      if (ginit.eq.0) then
+        ginit=1
+        call ffde_geom()
+      endif
+      idz=0
+      if (do_z_extra) idz=1
+      fu=1142
+      call ffde_open(fu,'q0')
+      write(fu) dble(itime),dble(ncyc),dble(idz),0d0
+      write(fu) dble(ncycxy)
+      write(fu) dble(nstepx)
+      write(fu) dble(nstepz_extra)
+      write(fu) dble(lminzij)
+      write(fu) dble(lmaxzij)
+      write(fu) mw_extra
+      write(fu) mus(1:im,1:jm,1:lm)
+      write(fu) mvs(1:im,1:jm,1:lm)
+      write(fu) mws(1:im,1:jm,1:lm)
+      write(fu) pv_south
+      write(fu) dble(ni_y)
+      write(fu) dble(ni_z)
+      n=0
+      do l=1,lm ; do j=1,jm ; n=n+ni_y(j,l) ; enddo ; enddo
+      write(fu) dble(n)
+      do l=1,lm ; do j=1,jm ; do k=1,ni_y(j,l)
+        write(fu) dble(j),dble(l),dble(i_y(k,j,l))
+      enddo ; enddo ; enddo
+      n=0
+      do l=1,lm ; do j=1,jm ; n=n+ni_z(j,l) ; enddo ; enddo
+      write(fu) dble(n)
+      do l=1,lm ; do j=1,jm ; do k=1,ni_z(j,l)
+        write(fu) dble(j),dble(l),dble(i_z(k,j,l))
+      enddo ; enddo ; enddo
+      close(fu)
+      end subroutine ffde_q0
+
+      subroutine ffde_ain(q,qmom)
+!@sum mass-unit state at AADVQ entry: ffd_qdyn_<itime>_ain.bin =
+!@+   [itime], RM=Q*MB (im,jm,lm), RMOM (9,im,jm,lm)
+      use model_com, only : itime
+      use resolution, only : im,jm,lm
+      implicit none
+      real*8 :: q(im,jm,lm),qmom(9,im,jm,lm)
+      integer :: fu
+      integer :: ffde_act
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      if (ffde_act.eq.0) return
+      fu=1143
+      call ffde_open(fu,'ain')
+      write(fu) dble(itime)
+      write(fu) q
+      write(fu) qmom
+      close(fu)
+      end subroutine ffde_ain
+
+      subroutine ffde_stg(stage,nc,ncxy,l,rm,rmom,mma)
+!@sum level-slice checkpoints of AADVQ into the ck stream (unit 1141):
+!@+   record = [stage,nc,ncxy,l], payload: stage 0 (after y checkflux,
+!@+   before AADVQY) RMOM(my),RMOM(myy) (im,jm each); stage 1 (after
+!@+   AADVQY), 2 (after AADVQX): RM (im,jm), RMOM (9,im,jm), MMA (im,jm);
+!@+   stage 3 (after z checkflux, before AADVQZ): RMOM(mz),RMOM(mzz).
+      use resolution, only : im,jm
+      implicit none
+      integer :: stage,nc,ncxy,l
+      real*8 :: rm(im,jm),rmom(9,im,jm),mma(im,jm)
+      integer :: fu
+      integer :: ffde_act
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      if (ffde_act.eq.0) return
+      fu=1141
+      write(fu) dble(stage),dble(nc),dble(ncxy),dble(l)
+      if (stage.eq.0) then
+        write(fu) rmom(2,:,:)
+        write(fu) rmom(5,:,:)
+      elseif (stage.eq.3) then
+        write(fu) rmom(3,:,:)
+        write(fu) rmom(6,:,:)
+      else
+        write(fu) rm
+        write(fu) rmom
+        write(fu) mma
+      endif
+      flush(fu)
+      end subroutine ffde_stg
+
+      subroutine ffde_carry(nc,l,mwdn,fdn,fdn0,fmomdn)
+!@sum stage 4: vertical carry after AADVQZ at iteration l:
+!@+   [4,nc,0,l], MWDN, FDN, FDN0 (im,jm), FMOMDN (9,im,jm)
+      use resolution, only : im,jm
+      implicit none
+      integer :: nc,l
+      real*8 :: mwdn(im,jm),fdn(im,jm),fdn0(im,jm),fmomdn(9,im,jm)
+      integer :: fu
+      integer :: ffde_act
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      if (ffde_act.eq.0) return
+      fu=1141
+      write(fu) 4d0,dble(nc),0d0,dble(l)
+      write(fu) mwdn
+      write(fu) fdn
+      write(fu) fdn0
+      write(fu) fmomdn
+      flush(fu)
+      end subroutine ffde_carry
+
+      subroutine ffde_full(stage,nc,rm,rmom,mma)
+!@sum stage 6 (after the L loop, before the extra z advection) and 7
+!@+   (after it): [stage,nc,0,0], RM (im,jm,lm), RMOM (9,im,jm,lm),
+!@+   MMA (im,jm,lm)
+      use resolution, only : im,jm,lm
+      implicit none
+      integer :: stage,nc
+      real*8 :: rm(im,jm,lm),rmom(9,im,jm,lm),mma(im,jm,lm)
+      integer :: fu
+      integer :: ffde_act
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      if (ffde_act.eq.0) return
+      fu=1141
+      write(fu) dble(stage),dble(nc),0d0,0d0
+      write(fu) rm
+      write(fu) rmom
+      write(fu) mma
+      flush(fu)
+      end subroutine ffde_full
+
+      subroutine ffde_out(rm,rmom,sbf,sbm,sfbm,scf,scm,sfcm,scf3d)
+!@sum AADVQ outputs: ffd_qdyn_<itime>_out.bin = [itime], RM (im,jm,lm),
+!@+   RMOM (9,im,jm,lm) mass units, MMA (im,jm,lm), then the zonal
+!@+   diagnostics SBF,SBM,SFBM,SCF,SCM,SFCM (jm,lm) and SCF3D
+!@+   (im,jm,lm).  Closes the ck stream.
+      use model_com, only : itime
+      use resolution, only : im,jm,lm
+      use atm_com, only : mma
+      implicit none
+      real*8 :: rm(im,jm,lm),rmom(9,im,jm,lm)
+      real*8 :: sbf(jm,lm),sbm(jm,lm),sfbm(jm,lm),scf(jm,lm)
+      real*8 :: scm(jm,lm),sfcm(jm,lm),scf3d(im,jm,lm)
+      integer :: fu
+      integer :: ffde_act
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      if (ffde_act.eq.0) return
+      close(1141)
+      fu=1143
+      call ffde_open(fu,'out')
+      write(fu) dble(itime)
+      write(fu) rm
+      write(fu) rmom
+      write(fu) mma(1:im,1:jm,1:lm)
+      write(fu) sbf
+      write(fu) sbm
+      write(fu) sfbm
+      write(fu) scf
+      write(fu) scm
+      write(fu) sfcm
+      write(fu) scf3d
+      close(fu)
+      end subroutine ffde_out
+
+      subroutine ffde_fin(q,qmom)
+!@sum final Q,QMOM (concentration units): ffd_qdyn_<itime>_fin.bin
+      use model_com, only : itime
+      use resolution, only : im,jm,lm
+      implicit none
+      real*8 :: q(im,jm,lm),qmom(9,im,jm,lm)
+      integer :: fu
+      integer :: ffde_act
+      real*8 :: ffde_sfac
+      character(len=3) :: ffde_comp
+      common /ffdest/ ffde_sfac,ffde_act,ffde_comp
+      save /ffdest/
+      if (ffde_act.eq.0) return
+      fu=1144
+      call ffde_open(fu,'fin')
+      write(fu) dble(itime)
+      write(fu) q
+      write(fu) qmom
+      close(fu)
+      ffde_act=0
+      end subroutine ffde_fin
+
       subroutine daily_atm(end_of_day)
       use filemanager
       use Dictionary_mod, only: sync_param
